### PR TITLE
Restructure extendLeaderNew() and its helpers

### PR DESCRIPTION
**Pull request by vasslitvinov**
* Restructure most of the "new" "Part 2" part of the implementation
of forall intents, specifically extendLeaderNew() and its helpers.
It still handles a variety of cases as before, now hopefully in
a more comprehensible manner.

* The following case is now handled nicely and uniformly
for all forall intents. The case is when the parallel iterator
(the one that is invoked by the forall loop)
itself contains a parallel loop (i.e. a forall or a for-loop
that invokes a leader or standalone iterator explicitly).

* The case from the previous bullet where the parallel loop is
a for-loop is handled by converting it to a forall loop first.
Handling a forall loop in this context is much simpler than
handling a for-loop, partly due to a better AST representation.
As a result, `eflopiMap` is no longer needed for the "new" implementation.

* The case when a yield in the parallel iterator is nested
inside 2 (or more) levels of parallel iterators has been causing INT_ASSERT.
Now it generates a "currently not implemented" error instead.

* ... except for the standalone iterator `walkdirs` in `FileSystem` module,
as tested by `modules/standard/FileSystem/filerator/bradc/findfiles-par`.

* To handle walkdirs and findfiles-par, recursive iterators are now supported
when the recursive invocation is made in a forall loop. (Previously it worked
only when such invocation was in a for-loop.) This works as long as there
are no reduce intents involved.

This PR's commits before squashing are 7278920..15b6a7e.

Testing:
- [x] linux64 --verify
- [x] gasnet multilocale tests
- [x] baseline

### CODE REVIEW

File: test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  coforall ooo in 1..3 {
+    yield 666;
+  }
+}
+
+iter pariter() {
+  yield 333;
+}
+
+iter pariter(param tag) {
```
Comment (mppf) [inline test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl]: Why doesn't this check tag== iterKind.standalone?

File: compiler/include/ForallStmt.h [hunk 2]
```diff
@@ -32,10 +32,14 @@ class ForallStmt : public Stmt
...
   bool       zippered()       const; // was 'zip' keyword used?
   AList&     inductionVariables();   // DefExprs, one per iterated expr
-  AList&     iteratedExpressions();  // SymExprs, CallExprs
+  AList&     iteratedExpressions();  // SymExprs, one per iterated expr
   AList&     intentVariables();      // DefExprs of LoopIntentVars
   BlockStmt* loopBody()       const; // the body of the forall loop
 
+  // when originating from a ForLoop
+  bool       iterCallAlreadyTagged() const;  // already has 'tag' actual
+  bool       noFindOuterVars()       const;  // no add'l shadow variables
```
Comment (mppf) [inline compiler/include/ForallStmt.h]: It seems odd to me that these two functions both just return fFromForLoop. Anyway, 'noFindOuterVars' seems like a strange method name; maybe it should be "outerVarsHandled" or something... Anyway both of these seem to have more to do with the flow of forall intents / iterator lowering than with ForallStmts fundamentally. For that reason it seems it might be better to just have an accessor for FromForLoop. Or, even better, avoid the need to even store 'fromForLoop' in this AST node at all, if it can become local information for some transformation pass.

File: compiler/include/expr.h [hunk 3]
```diff
@@ -474,7 +474,10 @@ static inline bool needsCapture(FnSymbol* taskFn) {
 }
 
 inline Symbol* ShadowVarSymbol::outerVarSym() const {
-  return this->outerVarSE()->symbol();
+  if (SymExpr* ovse = this->outerVarSE())
```
Comment (mppf) [inline compiler/include/expr.h]: It's not clear to me why you needed this change, but at least there's no reason for the assignment within the conditional. e.g.
```
SymExpr* ovse = this->outerVarSE();
if (ovse != NULL) return ovse->symbol();
else return NULL;
```

But it's obviously a style thing.

File: compiler/include/implementForallIntents.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                                                   Symbol* extraFormals[]);
+VarSymbol* localizeYieldForExtendLeader(Expr* origRetExpr, Expr* ref);
+void       checkAndRemoveOrigRetSym(Symbol* origRet, FnSymbol* parentFn);
+FnSymbol*  copyLeaderFn(FnSymbol* origFn, bool ignore_isResolved);
+
+void extendLeaderNew(ForallStmt* fs, FnSymbol* origIterFn, CallExpr* iterCall);
+
+static const char* astrArg(int ix, const char* base) {
+  return astr("x", istr(ix+1), "_", base);
+}
```
Comment (mppf) [inline compiler/include/implementForallIntents.h]: maybe name it something more to do with forall intents?

File: compiler/resolution/implementForallIntents2.cpp [hunk 5]
```diff
@@ -0,0 +1,1097 @@
...
+This file contains Part 2 of the new implementation of forall intents.
+"Part 2" is mostly about extending a copy of the parallel iterator in question
+with arguments and yield values corresponding to forall intents.
+
+For a reduce intent, the global reduce op instance needs to be
+properly allocated, deallocated and connected to the corresponding
+user outer variable. This is the job of "Part 1".
+
+For now, implementForallIntents2New() and immediately-adjacent functions
+are in implementForallIntents.cpp, for historical reasons.
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: Eh... why should you be conservative about moving code between the two files? If they fit better here, move them. My one "it would have been nice" request about this is that this PR seems to have a single commit combining code changes and code movement. That makes it much harder to follow. Try to move code in a separate commit.

File: compiler/resolution/implementForallIntents2.cpp [hunk 6]
```diff
@@ -0,0 +1,1097 @@
...
+  // Calling a function nested in the iterator?
+  bool         nested;
+
+  // The iterator's return symbol.
+  VarSymbol*   retSym;
+
+  // The FnSymbol for task startup/teardown code, usually 'curFn'.
+  // It is an enclosing function when descending into a task function
+  // that does not create a user-level task. Such a task function
+  // does not affect the intents.
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: I'm having trouble understanding this comment. wouldn't you want to add task startup/teardown code within the body of the task?

File: compiler/resolution/implementForallIntents2.cpp [hunk 7]
```diff
@@ -0,0 +1,1097 @@
...
+
+//
+// Is 'forLoop' a loop over a parallel iterator?
+// If so, fill in 'eInfo' with details.
+//
+static bool findCallToParallelIterator(ForLoop* forLoop, EflopiInfo& eInfo)
+{
+  Symbol* iterSym = forLoop->iteratorGet()->symbol();
+
+  // Find an assignment to 'iterSym'.
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: Is this usually located just a few statements before forLoop? If so, add a comment to that effect. If not, you might use for_SymbolSymExprs.

File: compiler/resolution/implementForallIntents2.cpp [hunk 8]
```diff
@@ -0,0 +1,1097 @@
...
+  Symbol* calltemp = toSymExpr(rhs1->get(1))->symbol();
+  CallExpr* asgnToCallTemp = NULL;
+  for (Expr* curr = asgnToIter->prev; curr; curr = curr->prev)
+    if (CallExpr* call = toCallExpr(curr))
+      if (call->isPrimitive(PRIM_MOVE))
+        if (SymExpr* lhs2 = toSymExpr(call->get(1)))
+          if (lhs2->symbol() == calltemp) {
+            asgnToCallTemp = call;
+            break;
+          }
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: To be clear, I don't like that we have to write this kind of pattern-matching-something-lowered code, but I don't have an alternative in mind. Would it help to add something-or-other to ForLoop? Or to promote the 'getIterator' idea to its own AST node or something?

File: compiler/resolution/implementForallIntents2.cpp [hunk 9]
```diff
@@ -0,0 +1,1097 @@
...
+  USR_PRINT(ctx.forall, "the forall loop that invokes the iterator is here");
+
+  // If we continue, we may get a const violation on x*_reduceShadowVar.
+  USR_STOP();
+}
+
+
+/////////// extendYieldNew, simple case ///////////
+
+static Symbol* tupcomForSerialYield(FIcontext& ctx, int ix)
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: let's expand out whatever you meant by 'tupcom'.

File: compiler/resolution/implementForallIntents2.cpp [hunk 10]
```diff
@@ -0,0 +1,1097 @@
...
+static ArgSymbol* newExtraFormal(FIcontext& ctx, FIinfo& fii, int ix,
+                                 Symbol* eActual)
+{
+    bool     isReduce   = fii.svSymbol->isReduce();
+    const char* eName   =
+      isReduce ? astrArg(ix, "reduceParent") :
+        ctx.nested ? eActual->name :
+          !strcmp(eActual->name, "_tuple_expand_tmp_") ?
+            astrArg(ix, "tet") :
+              astrArg(ix, eActual->name); // uniquify user arg name
```
Comment (mppf) [inline compiler/resolution/implementForallIntents2.cpp]: Would this be clearer as a series of conditionals? Not sure we're getting much by using the ternary operator here.

File: test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl (hunk 1 above)
Comment (vasslitvinov) [reply test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl]: To explain my thinking:  the check tag== iterKind.standalone is not needed in this simple test. In the sense that it should not affect the test outcome or the thing being tested. I omitted this check to shorted the code.

File: compiler/include/ForallStmt.h (hunk 2 above)
Comment (vasslitvinov) [reply compiler/include/ForallStmt.h]: I added a comment rationalizing the current design.

Replacing fFromForLoop with a HashSet is a good idea. I would like to defer that until later.

**Comment by vasslitvinov** [comment]
> I'd recommend putting everything back in to the original file for now, so that I can review this in a sane manner.

As Michael and I discussed off-PR, these changes are mostly "a complete rewrite". Therefore I did not follow this suggestion.

File: compiler/include/expr.h (hunk 3 above)
Comment (vasslitvinov) [reply compiler/include/expr.h]: This change is necessary because this->outerVarSE() can be NULL.

My reason for having a declaration within a conditional is that this style is quite common in our compiler. Now that I am used to it, I find it quite elegant.

File: compiler/include/ForallStmt.h (hunk 2 above)
Comment (vasslitvinov) [reply compiler/include/ForallStmt.h]: I renamed the second accessor to needToHandleOuterVars.

File: test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl (hunk 1 above)
Comment (mppf) [reply test/parallel/forall/vass/par-loops/2-enclosing-par-loops-error-4.chpl]: It seems to me that we should prefer easy to understand test cases over shorter test cases. They are not the same thing. Standalone iterators normally include the where clause.

**Comment by bradcray** [comment]
`implementForallIntents-2.cpp`?!?  We really couldn't come up with a more descriptive name for a file than that?  Is the intention that it will eventually replace `implementForallIntents.cpp`?

**Comment by mppf** [comment]
My personal viewpoint is that we need to actually remove the old implementation. I do not expect this to be so much work that we need to put it off another year. Then we can remove the `-2` suffix.

**Comment by vasslitvinov** [comment]
There used to be two phases with forall intents. 2 is for the second phase.
Like Michael says, the old implementation needs to go.
